### PR TITLE
Bi-directional DShot

### PR DESCRIPTION
**Pull request by andyp1per**
Add support for bi-directional dshot.

This is now working. To enable set the mask of bi-directional dshot channels
```
SERVO_BLH_BDMASK=15
```
For flight controllers with an IOMCU set:
```
SERVO_BLH_BDMASK=3840
```
Flying well on my KakuteF7Mini 3" quad - smooth and stable

See https://discuss.ardupilot.org/t/bi-directional-dshot-testers-wanted/64697 for the thread together with currently ported boards

### CODE REVIEW

**Comment by tridge** [comment]
good luck with this!

**Comment by andyp1per** [comment]
Squashed

**Review by tridge — CHANGES_REQUESTED** [review]
this is only a partial review, ran out of time

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 1]
```diff
@@ -1146,42 +1382,524 @@ void RCOutput::send_pulses_DMAR(pwm_group &group, uint32_t buffer_length)
...
+        value <<= len;
+        value |= 1 << (len - 1);
+        oldValue = buffer[i];
+        bits += len;
+    }
+    if (bits != 21) {
+        return 0xffff;
+    }
+
+    static const uint32_t decode[32] = {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1874]: this really needs an explanation :-)

**Comment by tridge** [comment]
a few general comments:
 - please test this with --enable-asserts
 - we need to look at the impact of this change on DMA mappings on current boards

**Comment by andyp1per** [comment]
Thanks for the feedback @tridge !
- I have done all the development with --enable-asserts
- My plan is to make the resolver feature triggered by HAL_BIDIR_DSHOT and then work through boards that people are willing to test to get the DMA mappings correct

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 2]
```diff
@@ -118,6 +118,13 @@ const AP_Param::GroupInfo AP_BLHeli::var_info[] = {
     // @User: Advanced
     AP_GROUPINFO("REMASK",  10, AP_BLHeli, channel_reversible_mask, 0),
 
+    // @Param: OPTS
+    // @DisplayName: BLHeli options
+    // @Description: Set of options that can be used to modify the BLHeli output.
+    // @Bitmask: 0:Bi-Dir DShot
+    // @User: Advanced
+    AP_GROUPINFO("OPTS",  11, AP_BLHeli, options, 0),
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_BLHeli/AP_BLHeli.cpp]: Added BDMASK

File: libraries/AP_BLHeli/AP_BLHeli.h [hunk 3]
```diff
@@ -91,6 +100,7 @@ class AP_BLHeli {
     AP_Int8 output_type;
     AP_Int8 control_port;
     AP_Int8 motor_poles;
+    AP_Int8 options;
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_BLHeli/AP_BLHeli.h]: Fixed

File: libraries/AP_HAL/RCOutput.h [hunk 4]
```diff
@@ -122,6 +122,12 @@ class AP_HAL::RCOutput {
      */
     virtual float    scale_esc_to_unity(uint16_t pwm) { return 0; }
 
+    /*
+      return the erpm and error rate for a channel if available
+     */
+    virtual uint16_t get_erpm(uint8_t chan) const { return 0; }
+    virtual float get_erpm_error_rate(uint8_t chan) const { return 0.0f; }
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL/RCOutput.h:129]: Added to logging

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:1812]: Added

File: libraries/AP_BLHeli/AP_BLHeli.cpp [hunk 5]
```diff
@@ -1352,18 +1359,32 @@ float AP_BLHeli::get_average_motor_frequency_hz() const
...
+            if (erpm < 0xFFFF) {
+                freqs[valid_escs++] = hal.rcout->get_erpm(i) * 200 / motor_poles;
+            }
+        } else if (last_telem[i].timestamp_ms && (now - last_telem[i].timestamp_ms < 1000)) {
             // slew the update
             const float slew = MIN(1.0f, (now - last_telem[i].timestamp_ms) * telem_rate / 1000.0f);
             freqs[valid_escs++] = (prev_motor_rpm[i] + (last_telem[i].rpm - prev_motor_rpm[i]) * slew) / 60.0f;
         }
     }
+    if (hal.util->get_soft_armed() && ocount++ % 200 == 0) {
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_BLHeli/AP_BLHeli.cpp]: Removed in favour of updating the existing debug and adding logging

**Comment by amilcarlucas** [comment]
How much memory does this cost for a CubeBlack? Will it work on the normal PWM1-8 outputs, or does it need to use the PWM_AUX pins?

**Comment by andyp1per** [comment]
Memory use unknown at this point, but not a lot - it will cost some flash. It needs to be on the AUX pins. Unlikely that we would do this for cube black - its EOL and I don't think @bugobliterator wants the support queries 😄

**Review by tridge — CHANGES_REQUESTED** [review]
i'd like a comparison of existing hwdef.h using configure_all.py

File: libraries/AP_HAL_ChibiOS/RCOutput.h [hunk 6]
```diff
@@ -217,6 +265,54 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput {
...
             thread_t *waiter;
         } serial;
+
+        // support for bi-directional dshot
+        uint16_t erpm[4];
+        volatile bool bidir_dshot_enabled;
+        volatile DshotState dshot_state;
+
+#if HAL_WITH_BIDIR_DSHOT
+        const stm32_dma_stream_t *ic_dma[4];
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.h]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.h [hunk 7]
```diff
@@ -279,6 +376,13 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput {
     // these values are for the local channels. Non-local channels are handled by IOMCU
     uint32_t en_mask;
     uint16_t period[max_channels];
+    // handling of bi-directional dshot
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.h:382]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.h [hunk 8]
```diff
@@ -279,6 +376,13 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput {
...
+    uint16_t _bidir_dshot_mask;
+    uint16_t _erpm[max_channels];
+    uint16_t _erpm_errors[max_channels] {};
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.h]: Removed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 9]
```diff
@@ -0,0 +1,569 @@
...
+        pwm_group &group = pwm_group_list[i];
+        if (((group.ch_mask << chan_offset) & mask) == 0) {
+            // this group is not affected
+            continue;
+        }
+        set_group_mode(group);
+    }
+}
+
+bool RCOutput::setup_group_ic_DMA(pwm_group &group)
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 10]
```diff
@@ -0,0 +1,569 @@
...
+        group.telem_rate[chan]++;
+    } else {
+        TOGGLE_PIN_DEBUG(57);
+        group.telem_err_rate[chan]++;
+        TOGGLE_PIN_DEBUG(57);
+    }
+
+    uint64_t now = AP_HAL::micros64();
+    if (chan == DEBUG_CHANNEL && (now  - group.last_print) > 1000000) {
+        hal.console->printf("TELEM: %d <%d Hz, %.1f%% err>", group.erpm[chan], group.telem_rate[chan],
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: its a bit too involved for that

File: libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp [hunk 11]
```diff
@@ -0,0 +1,569 @@
...
+{
+    switch (mode) {
+    case MODE_PWM_DSHOT150:
+        return 150000 * 5 / 4;
+    case MODE_PWM_DSHOT300:
+        return 300000 * 5 / 4;
+    case MODE_PWM_DSHOT600:
+        return 600000 * 5 / 4;
+    case MODE_PWM_DSHOT1200:
+        return 120000 * 5 / 4;
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput_bdshot.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py [hunk 12]
```diff
@@ -1492,7 +1499,27 @@ def write_PWM_config(f):
 # define HAL_PWM%u_DMA_CONFIG true, STM32_TIM_TIM%u_UP_DMA_STREAM, STM32_TIM_TIM%u_UP_DMA_CHAN
 #else
 # define HAL_PWM%u_DMA_CONFIG false, 0, 0
-#endif\n''' % (n, n, n, n, n, n))
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/chibios_hwdef.py:1500]: sid has fixed this

**Comment by andyp1per** [comment]
Flash cost is 3.7k
Memory cost is 1.3k

**Comment by tridge** [comment]
test plan:
 - [ ] believer with dshot telem (james, tridge, peter)
 - [ ] small quad without dshot
 - [ ] peter bixler
 - [ ] andy will fly quad with dshot telem

**Comment by tridge** [comment]
> Flash cost is 3.7k
> Memory cost is 1.3k

can you check these costs without bi-dshot enabled

**Comment by andyp1per** [comment]
So interestingly the flash cost is still 1.5k with this compiled out - so I need to figure out why that is.

**Comment by andyp1per** [comment]
Compiled out on a Pixracer works fine. Here is the log from a hover in my office: 
[00000286.zip](https://github.com/ArduPilot/ardupilot/files/5705129/00000286.zip)

**Comment by andyp1per** [comment]
I have compiled out a bit more so that BDMASK is no longer available unless the features is compiled in. Flash increase is down to 1k, which seems ok for the additional complexity that doing everything will entail - but can do it if necessary.

**Comment by andyp1per** [comment]
Did some hard flying on a 3" KakuteF7Min-based quad. All was perfect - no issues at all. Couple of logs below:
[bidir-20-12-17_10-47-34.zip](https://github.com/ArduPilot/ardupilot/files/5710322/bidir-20-12-17_10-47-34.zip)
[bidir-20-12-17_11-43-34.zip](https://github.com/ArduPilot/ardupilot/files/5710323/bidir-20-12-17_11-43-34.zip)

**Comment by andyp1per** [comment]
> test plan:
> 
>     * [ ]  believer with dshot telem (james, tridge, peter)
> 
>     * [X]  small quad without dshot
> 
>     * [ ]  peter bixler
> 
>     * [X]  andy will fly quad with dshot telem
> 
>     * [ ]  iomcu build test (likely will need to disable fport2 in iomcu code)

**Comment by andyp1per** [comment]
IOMCU on Durandal builds and runs fine with this change (I tested that there was output)
I have also now ported to a Pixracer quad and flown successfully with that

**Comment by andyp1per** [comment]
Did 12 flights today - 6 on the Kakute and 6 on the Pixracer. All worked fine no issues at all - can post logs if necessary

**Comment by andyp1per** [comment]
Rebuilt IOMCU and flew on my Y6B / Durandal with PWM/IOMCU outputs without issue

**Review by tridge — APPROVED** [review]

**Comment by davidbitton** [comment]
Does this addition pave the way towards RPM filtering?

**Comment by andyp1per** [comment]
@davidbitton we already support RPM filtering

**Comment by davidbitton** [comment]
@andyp1per thanks. I found some/the filtering code in `AP_BLHeli.cpp`.

**Comment by andyp1per** [comment]
image

**Pull request merged**